server.c

```diff
@@ -0,0 +1,259 @@
+    {
+      print_error ("Could not send data to client\n");
+      return (S_FAILURE);
+    }
+
+  int size;
+  if (recv_wrapper (socket_fd, &size, sizeof (int), 0) == S_FAILURE)
+    {
+      print_error ("Could not receive data from client\n");
+      return (S_FAILURE);
+    }
+
+  if (size != 0)
+    {
+      if (recv_wrapper (socket_fd, password, sizeof (password_t), 0)
+          == S_FAILURE)
+        {
+          print_error ("Could not receive data from client\n");
+          return (S_FAILURE);
+        }
+    }
+
+  return (S_SUCCESS);
+}
+
+static void *
+handle_client (void *arg)
+{
+  cl_context_t *cl_context = (cl_context_t *)arg;
+  cl_context_t local_context = *cl_context;
+  serv_context_t *context = local_context.context;
+
+  if (pthread_mutex_unlock (&cl_context->context->context.mutex) != 0)
+    {
+      print_error ("Could not unlock mutex\n");
+      return (NULL);
+    }
+
+  if (send_wrapper (local_context.socket_fd, context->context.config->hash,
+                    HASH_LENGTH, 0)
+      == S_FAILURE)
+    {
+      print_error ("Could not send hash to client\n");
+      return (NULL);
+    }
+
+  while (true)
+    {
+      task_t task;
+      if (queue_pop (&context->context.queue, &task) == S_FAILURE)
+        return (NULL);
+
+      task.to = task.from;
+      task.from = 0;
+
+      if (delegate_task (local_context.socket_fd, &task,
+                         context->context.password)
+          == S_FAILURE)
+        {
+          if (queue_push (&context->context.queue, &task) == S_FAILURE)
+            print_error ("Could not push to the queue\n");
+          return (NULL);
+        }
+
+      if (pthread_mutex_lock (&context->context.mutex) != 0)
+        {
+          print_error ("Could not lock a mutex\n");
+          return (NULL);
+        }
+      pthread_cleanup_push (cleanup_mutex_unlock, &context->context.mutex);
+
+      --context->context.passwords_remaining;
+      if (context->context.passwords_remaining == 0
+          || context->context.password[0] != 0)
+        if (pthread_cond_signal (&context->context.cond_sem) != 0)
+          {
+            print_error ("Could not signal a condition\n");
+            return (NULL);
+          }
+
+      pthread_cleanup_pop (!0);
+    }
+
+  close (context->socket_fd);
+
+  return (NULL);
+}
+
+static void *
+handle_clients (void *arg)
+{
+  serv_context_t *context = (serv_context_t *)arg;
+
+  while (true)
+    {
+      int client_socket = accept (context->socket_fd, NULL, NULL);
+      if (client_socket == -1)
+        {
+          print_error ("Could not accept new connection\n");
+          continue;
+        }
+
+      cl_context_t client_context = {
+        .context = context,
+        .socket_fd = client_socket,
+      };
+
+      if (thread_create (&context->context.thread_pool, handle_client,
+                         &client_context)
+          == S_FAILURE)
+        {
+          print_error ("Could not create client thread\n");
```

Comment:
in case of failure mutex need to be unlocked

Reply:
If thread_create fail, mutex remains locked and next lock will impose deadlock. Mutex needs to be unlocked in case of thread_create failure